alf/layers.py

```diff
@@ -424,6 +424,143 @@ def make_parallel(self, n):
+    member k in addition to the original FC weight matrix w. For input x, the
+    result for ensemble member k is calculated as :math:`(W \circ (s_k r_k^T)) x`.
+    This can be more efficiently calculated as :math:`(W (x \circ r_k)) \circ s_k`
+
+    """
+
+    def __init__(self,
+                 input_size,
+                 output_size,
+                 ensemble_size,
+                 output_ensemble_ids=True,
+                 activation=identity,
+                 use_bias=True,
+                 use_bn=False,
+                 use_ln=False,
+                 kernel_initializer=None,
+                 kernel_init_gain=1.0,
+                 bias_init_range=0.):
+        """
+        Args:
+            input_size (int): input size
+            output_size (int): output size
+            ensemble_size (int): ensemble size
+            output_ensemble_ids (bool): If True, the forward() funciton will return
+                a tuple of (result, ensemble_ids). If False, the forward() funciton
+                will return result only.
+            activation (Callable): activation function
+            use_bias (bool): whether use bias
+            use_bn (bool): whether use batch normalization.
+            use_ln (bool): whether use layer normalization
+            kernel_initializer (Callable): initializer for the FC layer kernel.
+                If none is provided a ``variance_scaling_initializer`` with gain as
+                ``kernel_init_gain`` will be used.
+            kernel_init_gain (float): a scaling factor (gain) applied to
+                the std of kernel init distribution. It will be ignored if
+                ``kernel_initializer`` is not None.
+            bias_init_range (float): biases are initialized uniformly in
+                [-bias_init_range, bias_init_range]
+        """
+        nn.Module.__init__(self)
+        self._r = nn.Parameter(torch.Tensor(ensemble_size, input_size))
+        self._s = nn.Parameter(torch.Tensor(ensemble_size, output_size))
+        self._ensemble_bias = nn.Parameter(
+            torch.Tensor(ensemble_size, output_size))
+        self._use_ensemble_bias = use_bias
+        self._ensemble_size = ensemble_size
+        self._output_ensemble_ids = output_ensemble_ids
+        self._bias_init_range = bias_init_range
+        super().__init__(
+            input_size,
+            output_size,
+            activation=identity,
+            use_bias=False,
+            use_bn=use_bn,
+            use_ln=use_ln,
+            kernel_initializer=kernel_initializer,
+            kernel_init_gain=kernel_init_gain)
+
+    def reset_parameters(self):
+        """Reinitialize parameters."""
+        super().reset_parameters()
+        torch.randint(
+            2, size=self._r.shape, dtype=torch.float32, out=self._r.data)
+        torch.randint(
+            2, size=self._s.shape, dtype=torch.float32, out=self._s.data)
+        self._r.data.mul_(2)
```

Comment:
Since ```torch.randint``` is already in [0, 2], do we still need ```mul_(2)```? 
If yes, it can be merged with randint anyway.

Same for ```self._s```.

It might be useful to comment on the initialization of the tuples of params for the low rank part somewhere, e.g. its range.